Update scientific python stack

bgruening opened the pull request:
This PR changes the way we distribute python packages in a substantial way. Testing and comments really appreciated.

* all python packages are now depending on one specific python version (2.7 in this case) using the `setup_python_environment`
* all `packages` have the python version in it's package name `_python_2_7_`
* packages based on python 3.4 can be created now
* you **can not** mix old python installations with the new `setup_python_environment` ones

@peterjc I have included a `reportlab` package, which can be used for biopython if you want to switch over to `setup_python_environment` for biopython as well

I think this makes installing packages easier and more consistent with the way we install R libraries and enables us to finally have python2 and python3 coexisting in the TS. Any objections?

peterjc commented:
This sounds good in principle - when tool authors using these Python 2.7 or 3.4 (or later 3.5 etc) dependencies should best practice be to use ``<command interpreter="python2.7">`` or ``<command interpreter="python3.4">`` to match? [This is future proof for Python 3.5 etc]

I presume given names like ``package_python_2_7_reportlab_3_1_44`` this would imply a future ``package_python_3_4_reportlab_3_1_44`` to match?

Assuming this goes in, I have no problem with us updating the Biopython packages. Would we keep ``package_biopython_1_65`` to mean Python 2 implicitly and add ``package_python_3_4_biopython_1_65``?

bgruening commented:
> This sounds good in principle - when tool authors using these Python 2.7 or 3.4 (or later 3.5 etc) dependencies should best practice be to use ``<command interpreter="python2.7">`` or ``<command interpreter="python3.4">`` to match? [This is future proof for Python 3.5 etc]

I don't think this is necessary. As soon as you require the package your default python will be the one you imported. 

> I presume given names like ``package_python_2_7_reportlab_3_1_44`` this would imply a future ``package_python_3_4_reportlab_3_1_44`` to match?

Yes, this is the idea.

> Assuming this goes in, I have no problem with us updating the Biopython packages. Would we keep ``package_biopython_1_65`` to mean Python 2 implicitly and add ``package_python_3_4_biopython_1_65``?

I would keep `package_biopython_1_65` as it is. It's the one with the old installation method and the new one should have the `_python_2_7_` included, indicating the new method.

peterjc commented:
>>  This sounds good in principle - when tool authors using these Python 2.7 or 3.4 (or later 3.5 etc) dependencies should best practice be to use ``<command interpreter="python2.7">`` or ``<command interpreter="python3.4">`` to match? [This is future proof for Python 3.5 etc]
>
> I don't think this is necessary. As soon as you require the package your default python will be the one you imported. 

OK, so just use ``<command interpreter="python">`` and declare via a ``<requirement ...>`` tag that it wants the package for Python 2.7 or 3.3 or ... - which would ensure ``$PATH`` includes that ``python`` (likely also available as ``python2.7`` or ``python3.3`` if you prefer making this explicit).

jmchilton commented:
Interesting... specifying an explicit Python version the way @peterjc suggests seems like it might be a good idea to me. It will make it more clear if something goes wrong during dependency resolution.  I feel like it adds more flexibility to the tools also by allowing more different versions of Python to be targeted without needing to lean on shed or custom dependency resolution. Hmm... on the other hand it does seem like the kind of idea that I like one day and then hate a year later - for reasons I cannot yet see now.

Also - at some point ``interpreter`` should be considered deprecated in lieu of ``$__tool_directory__`` - ``interpreter`` doesn't allow cheetah syntax or shell commands to appear before the script and so leads to frustrating, seemingly inconsistent behavior.

peterjc commented:
I think I would prefer to use ``<command interpreter="python3">`` or ``<command interpreter="python3.4">`` for a tool which doesn't work under Python 2. Otherwise as John says, if the dependencies are missing/failed to install, ``<command interpreter="python">`` would like get Python 2 and break.

@jmchilton could you clarify what you'd like to use instead of the ``<command interpreter="xxx">`` tag?

erasche commented on packages/package_python_2_7/tool_dependencies.xml:
```diff
@@ -3,39 +3,28 @@
     <repository name="package_xz_5_0_5" owner="iuc" prior_installation_required="True" />
   </package>
-
-  <package name="openssl" version="1.0.1">
```
python2.7 doesn't need SSL support, or ?

bgruening replied:
Need to look this up. Can be that I removed it because I originally wanted to not support ssl from the TS.

erasche replied:
ah, didn't see the + line below which changed to v1.0

erasche commented:
```
Linting repository /home/travis/build/galaxyproject/tools-iuc/packages/package_python_reportlab_3_1_44
Applying linter expansion... WARNING
.. WARNING: Failed to expand inclusions [{'source': '**'}]
Applying linter expected_files... WARNING
.. WARNING: Package repository does not contain a tool_dependencies.xml file.

Linting repository /home/travis/build/galaxyproject/tools-iuc/packages/package_python_pillow_2_7
Applying linter expansion... WARNING
.. WARNING: Failed to expand inclusions [{'source': '**'}]
Applying linter expected_files... WARNING
.. WARNING: Package repository does not contain a tool_dependencies.xml file.
```

bgruening commented:
@peterjc I usually prefer to not use the `interpreter` attribute. This causes a lot of inconsistency as @jmchilton already mentioned. As soon as you require a specific python version this should be the default version, as it is prepend to the PATH.  If you have python3.4 installed (system wide) I guess you can force to use this version with the `interpreter` attribute.

bgruening commented:
@erasche both directories are renamed, so this should not be an issue. They do not exists.

erasche commented:
@bgruening but they do exist, did you mean to remove the folder? There's still a `.shed.yml` there which is what's triggering the above error.

- https://github.com/galaxyproject/tools-iuc/tree/e0a7dbbca87d5359e8b92c78f230f5faf05308ad/packages/package_python_pillow_2_7
- https://github.com/galaxyproject/tools-iuc/tree/e0a7dbbca87d5359e8b92c78f230f5faf05308ad/packages/package_python_reportlab_3_1_44

bgruening commented:
@erasche @peterjc ready to be merged?

erasche commented:
@bgruening yeah, ready as far as I'm concerned...any resolution on whether or not you need SSL?

bgruening commented:
@erasche for the ssl libs in python you need them. I removed this dep because I thought we (as IUC) can not update it so regularly and that security relevant libraries should be maintained by the OS. 
We decided we can so I added them back.

erasche commented:
okay, yeah, looks fine as far as I'm concerned. I haven't tested.

bgruening commented:
I will wait a little bit and merge it tomorrow.

erasche commented:
Sorry for not being terribly helpful at the moment!
